src/coreComponents/physicsSolvers/wavePropagation/ElasticWaveEquationSEM.hpp

```diff
@@ -0,0 +1,426 @@
+
+#include "mesh/ExtrinsicMeshData.hpp"
+#include "physicsSolvers/SolverBase.hpp"
+#include "WaveSolverBase.hpp"
+
+
+
+namespace geosx
+{
+
+class ElasticWaveEquationSEM : public WaveSolverBase
+{
+public:
+
+  using EXEC_POLICY = parallelDevicePolicy< 32 >;
+  using ATOMIC_POLICY = parallelDeviceAtomic;
+
+  /**
+   * @brief Safeguard for timeStep. Used to avoid memory issue due to too small value.
+   */
+  static constexpr real64 epsilonLoc = 1e-8;
+
+  ElasticWaveEquationSEM( const std::string & name,
+                          Group * const parent );
+
+  virtual ~ElasticWaveEquationSEM() override;
+
+  ElasticWaveEquationSEM() = delete;
+  ElasticWaveEquationSEM( ElasticWaveEquationSEM const & ) = delete;
+  ElasticWaveEquationSEM( ElasticWaveEquationSEM && ) = default;
+
+  ElasticWaveEquationSEM & operator=( ElasticWaveEquationSEM const & ) = delete;
+  ElasticWaveEquationSEM & operator=( ElasticWaveEquationSEM && ) = delete;
+
+
+  static string catalogName() { return "ElasticSEM"; }
+
+  virtual void initializePreSubGroups() override;
+
+  virtual void registerDataOnMesh( Group & meshBodies ) override final;
+
+
+  /**
+   * @defgroup Solver Interface Functions
+   *
+   * These functions provide the primary interface that is required for derived classes
+   */
+  /**@{*/
+  virtual
+  real64 solverStep( real64 const & time_n,
+                     real64 const & dt,
+                     integer const cycleNumber,
+                     DomainPartition & domain ) override;
+
+  virtual
+  real64 explicitStep( real64 const & time_n,
+                       real64 const & dt,
+                       integer const cycleNumber,
+                       DomainPartition & domain ) override;
+
+  /**
+   * @brief Multiply the precomputed term by the Ricker and add to the right-hand side
+   * @param cycleNumber the cycle number/step number of evaluation of the source
+   * @param rhsx the right hand side vector to be computed (x-component)
+   * @param rhsy the right hand side vector to be computed (x-component)
+   * @param rhsz the right hand side vector to be computed (x-component)
+   */
+  void addSourceToRightHandSide( integer const & cycleNumber, arrayView1d< real64 > const rhsx, arrayView1d< real64 > const rhsy, arrayView1d< real64 > const rhsz );
+
+  /**
+   * TODO: move implementation into WaveSolverBase
+   * @brief Compute the sesimic traces for a given variable at each receiver coordinate at a given time, using the field values at the
+   * last two timesteps.
+   * @param time_n the time corresponding to the field values at iteration n
+   * @param dt the simulation timestep
+   * @param timeSeismo the time at which the seismogram is computed
+   * @param iSeismo the index of the seismogram time in the seismogram array
+   * @param var_np1 the field values at time_n + dt
+   * @param var_n the field values at time_n
+   * @param varAtreceivers the array holding the trace values, where the output is written
+   */
```

Comment:
It is just a small detail, but you don't have to copy-paste the Doxygen documentation for functions that are overriding base class functions, because Doxygen figures things out automatically if you only document the base class functions.

Reply:
Okay noted, than you !